File: 0046-permutations/0046-permutations.java
class Solution {
    List<Integer> path = new LinkedList<>();
    List<List<Integer>> ans = new ArrayList<>();
    boolean[] used;
    
    public List<List<Integer>> permute(int[] nums) {
        used = new boolean[nums.length];
        backtracking(nums);
        return ans;
    }
    
    private void backtracking(int[] nums) {
        if (path.size() == nums.length) {
            ans.add(new ArrayList<>(path));
            return;
        }
        
        for (int i = 0; i < nums.length; i++) {
            if (used[i] == true) {
                continue;
            }
            
            path.add(nums[i]);
            used[i] = true;
            backtracking(nums);
            path.removeLast();
            used[i] = false;      
        }
    }
}

File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        ans = []

        def backtracking(path: List[int], used: set[int]) -> None:
            if len(path) == len(nums):
                ans.append(path[:])

            for i in range(0, len(nums)):
                if i in used:
                    continue

                path.append(nums[i])
                used.add(i)
                backtracking(path, used)
                used.remove(i)
                path.pop()
        
        backtracking([], set())
        return ans
